bob-apps/bin/bobp.rs

```diff
@@ -254,92 +267,199 @@ impl Debug for TaskConfig {
-        self.exist_time_ns_st
-            .fetch_add(duration.as_nanos() as u64, Ordering::Relaxed);
-        self.exist_count_st.fetch_add(1, Ordering::Relaxed);
+    fn process_and_print(&mut self) -> f64 {
+        let bit = self.common.current_bit();
+        let spd = self.size.get_diff() as f64 / 1024.0 / self.period_s;
+        println!("exist: {:>6} rps | err {:5} | {:>6.2} kb/s | lat {:>6.2} ms", 
+            bit.count_spd, bit.error, spd, bit.latency);
+        spd
     }
+}
 
-    fn save_get_error(&self, status: Status) {
-        let mut guard = self.get_errors.lock().expect("mutex");
-        guard
-            .entry(status.code() as CodeRepresentation)
-            .and_modify(|i| *i += 1)
-            .or_insert(1);
+type CodeRepresentation = i32; // Because tonic::Code does not implement hash
 
-        self.get_error_count.fetch_add(1, Ordering::SeqCst);
-        debug!("{}", status.message())
-    }
+#[derive(Default)]
+struct OperationStatistics {
+    total: AtomicU64,
+    error_count: AtomicU64,
+    time_ns_st: AtomicU64,
+    count_st: AtomicU64,
+    errors: Mutex<HashMap<CodeRepresentation, u64>>,
+}
 
-    fn save_put_error(&self, status: Status) {
-        let mut guard = self.put_errors.lock().expect("mutex");
+impl OperationStatistics {
+    async fn save_error(&self, status: Status) {
+        let mut guard = self.errors.lock().expect("mutex");
         guard
             .entry(status.code() as CodeRepresentation)
             .and_modify(|i| *i += 1)
             .or_insert(1);
 
-        self.put_error_count.fetch_add(1, Ordering::SeqCst);
+        self.error_count.fetch_add(1, Ordering::SeqCst);
         debug!("{}", status.message())
     }
 
-    fn save_exist_error(&self, status: Status) {
-        let mut guard = self.exist_errors.lock().expect("mutex");
-        guard
-            .entry(status.code() as CodeRepresentation)
-            .and_modify(|i| *i += 1)
-            .or_insert(1);
+    fn save_single_thread_time(&self, duration: &Duration) {
+        self.time_ns_st
+            .fetch_add(duration.as_nanos() as u64, Ordering::Relaxed);
+        self.count_st.fetch_add(1, Ordering::Relaxed);
+    }    
+}
 
-        self.exist_error_count.fetch_add(1, Ordering::SeqCst);
-        debug!("{}", status.message())
-    }
+#[derive(Default)]
+struct PutStatistics {
+    common: OperationStatistics,
+    verified: AtomicU64,
+}
+
+#[derive(Default)]
+struct GetStatistics {
+    common: OperationStatistics,
+    size_bytes: AtomicU64,
+}
+
+#[derive(Default)]
+struct ExistStatistics {
+    common: OperationStatistics,
+    size_bytes: AtomicU64,
+    presented_keys: AtomicU64,
+}
+
+#[derive(Default)]
+struct DeleteStatistics {
+    common: OperationStatistics,
+    verified: AtomicU64,
+}
+
+#[derive(Default)]
+struct Statistics {
+    put: PutStatistics,
+    get: GetStatistics,
+    exist: ExistStatistics,
+    delete: DeleteStatistics,
```

Comment:
The struct is large by itself, so I think it is better to wrap every item here with `Arc<>`